Client/Frontend/Home/PanelDataObservers.swift

```diff
@@ -48,28 +50,33 @@ class ActivityStreamDataObserver: DataObserver {
         events.forEach { NotificationCenter.default.removeObserver(self, name: $0, object: nil) }
     }
     
-    func invalidate(highlights: Bool) {
+    func refresh(highlights: Bool) {
+        guard !profile.isShutdown else {
+            return
+        }
         self.delegate?.willInvalidateDataSources()
 
-        let notify = {
+        // Highlights are cached for 15 mins
+        let invalidateHighlights = highlights ? true : (Date.now() - lastInvalidation > invalidationTime)
+        lastInvalidation = invalidateHighlights ? Date.now() : lastInvalidation
+
+        let topSitesIsValid = profile.prefs.boolForKey(PrefsKeys.KeyTopSitesCacheIsValid) ?? false
+
+        self.profile.recommendations.repopulateAll(!topSitesIsValid, invalidateHighlights: invalidateHighlights).uponQueue(.main) { _ in
+            self.profile.prefs.setBool(true, forKey: PrefsKeys.KeyTopSitesCacheIsValid)
             self.delegate?.didInvalidateDataSources()
         }
-        
-        let invalidateTopSites: () -> Success = {
-            self.profile.history.setTopSitesNeedsInvalidation()
-            return self.profile.history.updateTopSitesCacheIfInvalidated() >>> succeed
-        }
+    }
 
-        let shouldInvalidate = highlights ? true : (Date.now() - lastInvalidation > invalidationTime)
-        lastInvalidation = shouldInvalidate ? Date.now() : lastInvalidation
-        let query = shouldInvalidate ? [self.profile.recommendations.invalidateHighlights, invalidateTopSites] : [invalidateTopSites]
-        accumulate(query) >>> effect(notify)
+    func invalidateTopSites() {
+         profile.prefs.setBool(false, forKey: PrefsKeys.KeyTopSitesCacheIsValid)
```

Comment:
Invalidating topsites is now explicit. AS will flip the flag if topsites needs to be invalidated. This prevents needless topsite invalidations.